# Data Aggregation using Pandas

In [3]:
# import the pandas library
import pandas as pd

In [4]:
# read in the data
safi = pd.read_csv('data/SAFI_clean.csv')

In [5]:
# use describe() to get summary statistics of the data
safi.describe()

key_ID  no_membrs   years_liv       rooms   liv_count    no_meals
count  131.000000  131.00000  131.000000  131.000000  131.000000  131.000000
mean    85.473282    7.19084   23.053435    1.740458    2.366412    2.603053
std     63.151628    3.17227   16.913041    1.092547    1.082775    0.491143
min      1.000000    2.00000    1.000000    1.000000    1.000000    2.000000
25%     32.500000    5.00000   12.000000    1.000000    1.000000    2.000000
50%     66.000000    7.00000   20.000000    1.000000    2.000000    3.000000
75%    138.000000    9.00000   27.500000    2.000000    3.000000    3.000000
max    202.000000   19.00000   96.000000    8.000000    5.000000    3.000000

In [6]:
safi.max()

key_ID                                                        202
village                                                     Ruaca
interview_date                               2017-06-04T00:00:00Z
no_membrs                                                      19
years_liv                                                      96
respondent_wall_type                                    sunbricks
rooms                                                           8
liv_count                                                       5
no_meals                                                        3
months_lack_food                                             none
instanceID              uuid:ffc83162-ff24-4a87-8709-eff17abc0b3b
dtype: object

In [7]:
safi['rooms'].mean()

1.7404580152671756

### What to do with categorial/text data?

In [8]:
# get unique entries of a variable
safi['village'].unique()

array(['God', 'Chirodzo', 'Ruaca'], dtype=object)

In [9]:
safi['village'].count()

131

# %load code/counts
1. Get the unique entries for the `affect_conflicts` variable
2. Get the count value for `affect_conflicts` and compare it with the count for `respondent_wall_type`.
3. Why do you think they are different?

### Solutions to exercise

In [15]:
# 1.
safi['affect_conflicts'].unique()

array([nan, 'once', 'never', 'more_once', 'frequently'], dtype=object)

In [16]:
# 2.
safi['affect_conflicts'].count()

92

3. the count for 'affect_conflicts' is 92 and the count for 'villages' or 'respondent_wall_type' was 131. The reason for the difference is: there are missing values (indicated by NaN) in 'affect_conflicts' which are not counted (see below)

In [17]:
safi['affect_conflicts']

0             NaN
1            once
2             NaN
3             NaN
4             NaN
5             NaN
6           never
7           never
8           never
9           never
10            NaN
11          never
12          never
13            NaN
14           once
15            NaN
16            NaN
17            NaN
18            NaN
19            NaN
20          never
21            NaN
22            NaN
23          never
24          never
25          never
26            NaN
27      more_once
28     frequently
29            NaN
          ...    
101    frequently
102         never
103           NaN
104         never
105         never
106           NaN
107     more_once
108         never
109           NaN
110         never
111           NaN
112           NaN
113    frequently
114    frequently
115         never
116         never
117    frequently
118         never
119         never
120         never
121         never
122         never
123         never
124         never
125       

In [18]:
# test DataFram for missing values --> False = cell is not a missing value, True = cell is a missing value
safi.isna()

key_ID  village  interview_date  no_membrs  years_liv  \
0     False    False           False      False      False   
1     False    False           False      False      False   
2     False    False           False      False      False   
3     False    False           False      False      False   
4     False    False           False      False      False   
5     False    False           False      False      False   
6     False    False           False      False      False   
7     False    False           False      False      False   
8     False    False           False      False      False   
9     False    False           False      False      False   
10    False    False           False      False      False   
11    False    False           False      False      False   
12    False    False           False      False      False   
13    False    False           False      False      False   
14    False    False           False      False      False   
15    False    False           False      False      False   
16    False    False           False      False      False   
17    False    False           False      False      False   
18    False    False           False      False      False   
19    False    False           False      False      False   
20    False    False           False      False      False   
21    False    False           False      False      False   
22    False    False           False      False      False   
23    False    False           False      False      False   
24    False    False           False      False      False   
25    False    False           False      False      False   
26    False    False           False      False      False   
27    False    False           False      False      False   
28    False    False           False      False      False   
29    False    False           False      False      False   
..      ...      ...             ...        ...        ...   
101   False    False           False      False      False   
102   False    False           False      False      False   
103   False    False           False      False      False   
104   False    False           False      False      False   
105   False    False           False      False      False   
106   False    False           False      False      False   
107   False    False           False      False      False   
108   False    False           False      False      False   
109   False    False           False      False      False   
110   False    False           False      False      False   
111   False    False           False      False      False   
112   False    False           False      False      False   
113   False    False           False      False      False   
114   False    False           False      False      False   
115   False    False           False      False      False   
116   False    False           False      False      False   
117   False    False           False      False      False   
118   False    False           False      False      False   
119   False    False           False      False      False   
120   False    False           False      False      False   
121   False    False           False      False      False   
122   False    False           False      False      False   
123   False    False           False      False      False   
124   False    False           False      False      False   
125   False    False           False      False      False   
126   False    False           False      False      False   
127   False    False           False      False      False   
128   False    False           False      False      False   
129   False    False           False      False      False   
130   False    False           False      False      False   

     respondent_wall_type  rooms  memb_assoc  affect_conflicts  liv_count  \
0                   False  False        True              True      False   
1     

In [19]:
# get the count of missing values for each column in safi DataFrame by summing over each column 
# --> False is represented by 0 and True by 1 in python
safi.isna().sum()

key_ID                   0
village                  0
interview_date           0
no_membrs                0
years_liv                0
respondent_wall_type     0
rooms                    0
memb_assoc              39
affect_conflicts        39
liv_count                0
items_owned             10
no_meals                 0
months_lack_food         0
instanceID               0
dtype: int64

In [20]:
# remove all rows where any of the columns has a missing value
safi.dropna()

key_ID   village        interview_date  no_membrs  years_liv  \
1         1       God  2016-11-17T00:00:00Z          7          9   
6         7       God  2016-11-17T00:00:00Z          6         38   
7         8  Chirodzo  2016-11-16T00:00:00Z         12         70   
8         9  Chirodzo  2016-11-16T00:00:00Z          8          6   
9        10  Chirodzo  2016-12-16T00:00:00Z         12         23   
11       12       God  2016-11-21T00:00:00Z          7         20   
12       13       God  2016-11-21T00:00:00Z          6          8   
14       15       God  2016-11-21T00:00:00Z          5         30   
23       24     Ruaca  2016-11-21T00:00:00Z          6          4   
24       25     Ruaca  2016-11-21T00:00:00Z         11          6   
25       26     Ruaca  2016-11-21T00:00:00Z          3         20   
28       29     Ruaca  2016-11-21T00:00:00Z          7         10   
31       32     Ruaca  2016-11-21T00:00:00Z         19         69   
32       33     Ruaca  2016-11-21T00:00:00Z          8         34   
33       34  Chirodzo  2016-11-17T00:00:00Z          8         18   
34       35  Chirodzo  2016-11-17T00:00:00Z          5         45   
35       36  Chirodzo  2016-11-17T00:00:00Z          6         23   
37       38       God  2016-11-17T00:00:00Z         10         19   
39       40       God  2016-11-17T00:00:00Z          9         23   
41       42       God  2016-11-17T00:00:00Z          8          8   
42       43  Chirodzo  2016-11-17T00:00:00Z          7         29   
44       45  Chirodzo  2016-11-17T00:00:00Z          9          7   
45       46  Chirodzo  2016-11-17T00:00:00Z         10         42   
46       47  Chirodzo  2016-11-17T00:00:00Z          2          2   
49       50  Chirodzo  2016-11-16T00:00:00Z          6          7   
51       52  Chirodzo  2016-11-16T00:00:00Z         11         15   
52       21  Chirodzo  2016-11-16T00:00:00Z          8         16   
55       56  Chirodzo  2016-11-16T00:00:00Z         12         23   
56       57  Chirodzo  2016-11-16T00:00:00Z          4         27   
57       58  Chirodzo  2016-11-16T00:00:00Z         11         45   
..      ...       ...                   ...        ...        ...   
95      101       God  2017-04-27T00:00:00Z          3          4   
96      103     Ruaca  2017-04-27T00:00:00Z          6         96   
97      102     Ruaca  2017-04-28T00:00:00Z         12         15   
98       78     Ruaca  2017-04-28T00:00:00Z          6         48   
99       80     Ruaca  2017-04-28T00:00:00Z          5         12   
100     104     Ruaca  2017-04-28T00:00:00Z         14         52   
101     105     Ruaca  2017-04-28T00:00:00Z          6         40   
102     106       God  2017-04-30T00:00:00Z         15         22   
104     110     Ruaca  2017-05-03T00:00:00Z          6         22   
105     113     Ruaca  2017-05-03T00:00:00Z         11         26   
107     125     Ruaca  2017-05-04T00:00:00Z          5         14   
108     119     Ruaca  2017-05-04T00:00:00Z          3         14   
110     108       God  2017-05-11T00:00:00Z         15         22   
113     144     Ruaca  2017-05-18T00:00:00Z          7          5   
114     143     Ruaca  2017-05-18T00:00:00Z         10         24   
115     150     Ruaca  2017-05-18T00:00:00Z          7          8   
116     159       God  2017-05-18T00:00:00Z          4         24   
117     160       God  2017-06-03T00:00:00Z          7         13   
118     165     Ruaca  2017-06-03T00:00:00Z          9         14   
119     166     Ruaca  2017-06-03T00:00:00Z         11         16   
120     167     Ruaca  2017-06-03T00:00:00Z          8         24   
121     174     Ruaca  2017-06-03T00:00:00Z         12         25   
122     175     Ruaca  2017-06-03T00:00:00Z          7         36   
123     189     Ruaca  2017-06-03T00:00:00Z         15         16   
124     191     Ruaca  2017-06-03T00:00:00Z         10          5   
125     192  Chirodzo  2017-06-03T00:00:00Z          9         20   
126

In [21]:
# create a grouped_by object in which our data are grouped by 'village'
grouped_by = safi.groupby("village")

In [22]:
# then operate count on the grouped object --> returns the count for column per village
grouped_by.count()

key_ID  interview_date  no_membrs  years_liv  respondent_wall_type  \
village                                                                        
Chirodzo      39              39         39         39                    39   
God           43              43         43         43                    43   
Ruaca         49              49         49         49                    49   

          rooms  memb_assoc  affect_conflicts  liv_count  items_owned  \
village                                                                 
Chirodzo     39          27                27         39           36   
God          43          25                25         43           39   
Ruaca        49          40                40         49           46   

          no_meals  months_lack_food  instanceID  
village                                           
Chirodzo        39                39          39  
God             43                43          43  
Ruaca           49                49          49

In [23]:
# create an object no_members that holds the mean value of 'no_membrs' calculated per 'village' ad 'affect_conflicts'
no_membrs = safi.groupby(['village','affect_conflicts']).mean()['no_membrs']

In [18]:
no_membrs

village   affect_conflicts
Chirodzo  frequently           9.000000
          more_once            6.875000
          never                8.916667
          once                 7.000000
God       frequently           7.000000
          more_once            8.000000
          never                7.428571
          once                 6.000000
Ruaca     frequently           7.833333
          more_once            7.307692
          never                8.150000
          once                10.000000
Name: no_membrs, dtype: float64

# %load code/groupby
1. Read in the SAFI_clean.csv dataset.
2. Get a list of the different `rooms` values.
3. Groupby `rooms` and describe the results.
4. Remove all rows with NaN values.
5. repeat steps 2 & 3 and compare the results.

### Solution to exercise

In [24]:
# 1.
safi = pd.read_csv('data/SAFI_clean.csv')

In [25]:
# 2. 
safi['rooms'].unique()

array([1, 3, 5, 2, 4, 8])

In [27]:
# 3.
safi.groupby('rooms').describe()

key_ID                                                            \
       count        mean        std    min    25%    50%    75%    max   
rooms                                                                    
1       73.0   78.958904  61.534416    1.0   33.0   61.0  125.0  198.0   
2       33.0  100.515152  66.276938   15.0   46.0   76.0  174.0  201.0   
3       17.0   70.352941  55.865845    8.0   25.0   59.0  110.0  193.0   
4        5.0  135.400000  71.716804   23.0  117.0  144.0  191.0  202.0   
5        2.0   58.000000  67.882251   10.0   34.0   58.0   82.0  106.0   
8        1.0  127.000000        NaN  127.0  127.0  127.0  127.0  127.0   

      liv_count            ...  no_membrs       years_liv             \
          count      mean  ...        75%   max     count       mean   
rooms                      ...                                         
1          73.0  2.068493  ...       7.00  15.0      73.0  21.095890   
2          33.0  2.848485  ...      10.00  19.0      33.0  26.181818   
3          17.0  2.823529  ...      11.00  12.0      17.0  28.411765   
4           5.0  2.400000  ...      10.00  12.0       5.0  14.000000   
5           2.0  2.000000  ...      14.25  15.0       2.0  22.500000   
8           1.0  1.000000  ...       4.00   4.0       1.0  18.000000   

                                                  
             std   min    25%   50%    75%   max  
rooms                                             
1      17.158707   1.0   8.00  16.0  27.00  96.0  
2      16.602813   4.0  16.00  23.0  30.00  79.0  
3      18.100617   6.0  18.00  21.0  37.00  70.0  
4       9.974969   5.0   5.00  12.0  20.00  28.0  
5       0.707107  22.0  22.25  22.5  22.75  23.0  
8            NaN  18.0  18.00  18.0  18.00  18.0  

[6 rows x 40 columns]

In [28]:
# 4. and 5.
safi.dropna().groupby('rooms').describe()

key_ID                                                             \
       count        mean        std    min     25%    50%    75%    max   
rooms                                                                     
1       44.0   98.295455  60.959605    1.0   44.50   84.0  160.5  195.0   
2       25.0  104.840000  63.948339   15.0   56.00   89.0  174.0  199.0   
3       14.0   71.928571  58.515566    8.0   27.25   61.5   99.5  193.0   
4        3.0  179.000000  30.805844  144.0  167.50  191.0  196.5  202.0   
5        2.0   58.000000  67.882251   10.0   34.00   58.0   82.0  106.0   

      liv_count            ...  no_membrs       years_liv             \
          count      mean  ...        75%   max     count       mean   
rooms                      ...                                         
1          44.0  2.318182  ...       8.25  15.0      44.0  23.000000   
2          25.0  3.080000  ...      11.00  19.0      25.0  27.680000   
3          14.0  2.785714  ...      11.00  12.0      14.0  27.000000   
4           3.0  2.666667  ...      11.00  12.0       3.0   7.333333   
5           2.0  2.000000  ...      14.25  15.0       2.0  22.500000   

                                                  
             std   min    25%   50%    75%   max  
rooms                                             
1      17.885944   2.0  11.25  17.5  30.25  96.0  
2      18.350114   4.0  15.00  24.0  31.00  79.0  
3      17.832555   6.0  16.50  20.5  34.25  70.0  
4       4.041452   5.0   5.00   5.0   8.50  12.0  
5       0.707107  22.0  22.25  22.5  22.75  23.0  

[5 rows x 40 columns]

### explanation:
- by dropping all missing data we remove the rows where rooms is 8 and the summary statistics for the variables is different to our original data (the one including missing data)